SingularityService/src/main/java/com/hubspot/singularity/scheduler/SingularityUsagePoller.java

```diff
@@ -183,12 +211,48 @@ public void runActionOnPoll() {
             if (isLongRunning(task) ||  isConsideredLongRunning(task)) {
               updateLongRunningTasksUsage(longRunningTasksUsage, latestUsage.getMemoryTotalBytes(), taskCpusUsed, latestUsage.getDiskTotalBytes());
             }
-            SingularityTaskCurrentUsage currentUsage = new SingularityTaskCurrentUsage(latestUsage.getMemoryTotalBytes(), now, taskCpusUsed, latestUsage.getDiskTotalBytes());
+            currentUsage = new SingularityTaskCurrentUsage(latestUsage.getMemoryTotalBytes(), now, taskCpusUsed, latestUsage.getDiskTotalBytes());
 
             usageManager.saveCurrentTaskUsage(taskId, currentUsage);
 
             cpusUsedOnSlave += taskCpusUsed;
           }
+
+          if (slaveOverloaded && configuration.isShuffleTasksForOverloadedSlaves() && currentUsage != null && currentUsage.getCpusUsed() > 0) {
+            if (isLongRunning(task)) {
+              Optional<SingularityTaskHistoryUpdate> maybeCleanupUpdate = taskManager.getTaskHistoryUpdate(task, ExtendedTaskState.TASK_CLEANING);
+              if (maybeCleanupUpdate.isPresent() && isTaskShuffleCleanup(maybeCleanupUpdate.get())) {
+                LOG.trace("Task {} already being cleaned up to spread cpu usage, skipping", taskId);
+                shuffledTasks++;
+              } else {
+                possibleTasksToShuffle.add(new TaskIdWithUsage(task, currentUsage));
+              }
+            }
+          }
+        }
+
+        if (slaveOverloaded && configuration.isShuffleTasksForOverloadedSlaves()) {
+          possibleTasksToShuffle.sort((u1, u2) -> Double.compare(u2.getUsage().getCpusUsed(), u1.getUsage().getCpusUsed()));
+          for (TaskIdWithUsage taskIdWithUsage : possibleTasksToShuffle) {
+            if (cpuOverage <= 0 || shuffledTasks > configuration.getMaxTasksToShuffleForCpuOverage()) {
```

Comment:
if this line is reached then, `slaveOverloaded` must be `true`. Since `slaveOverloaded` is only true if `used > total`, then it should be an error if `cpuOverage <= 0`. If you want to keep the check, it might be worth putting in an error log

Reply:
cpuOverage is mutable here inside the for loop. We decrement it for each task that we move off of the slave and this check breaks us out if we either removed enough cpu, or hit our max number of tasks to shuffle at once

Reply:
ah, i see the update now. my mistake